pkg/mixin/feed/generate.go

```diff
@@ -0,0 +1,168 @@
+package feed
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"sort"
+	"time"
+
+	"github.com/cbroglie/mustache"
+	"github.com/deislabs/porter/pkg/context"
+	"github.com/pkg/errors"
+)
+
+type GenerateOptions struct {
+	SearchDirectory string
+	AtomFile        string
+	TemplateFile    string
+}
+
+func (o *GenerateOptions) Validate(c *context.Context) error {
+	err := o.ValidateSearchDirectory(c)
+	if err != nil {
+		return err
+	}
+
+	return o.ValidateTemplateFile(c)
+}
+
+func (o *GenerateOptions) ValidateSearchDirectory(cxt *context.Context) error {
+	if o.SearchDirectory == "" {
+		wd, err := os.Getwd()
+		if err != nil {
+			return errors.Wrap(err, "could not get current working directory")
+		}
+
+		o.SearchDirectory = wd
+	}
+
+	if _, err := cxt.FileSystem.Stat(o.SearchDirectory); err != nil {
+		return errors.Wrapf(err, "invalid --directory %s", o.SearchDirectory)
+	}
+
+	return nil
+}
+
+func (o *GenerateOptions) ValidateTemplateFile(cxt *context.Context) error {
+	if _, err := cxt.FileSystem.Stat(o.TemplateFile); err != nil {
+		return errors.Wrapf(err, "invalid --template %s", o.TemplateFile)
+	}
+
+	return nil
+}
+
+type mixinFileset struct {
+	Mixin   string
+	Version string
+	Files   []mixinFile
+}
+
+func (f *mixinFileset) Updated() string {
+	return toAtomTimestamp(f.GetLastUpdated())
+}
+
+func (f *mixinFileset) GetLastUpdated() time.Time {
+	var max time.Time
+	for _, f := range f.Files {
+		if f.Updated.After(max) {
+			max = f.Updated
+		}
+	}
+	return max
+}
+
+type mixinEntries []*mixinFileset
+
+func (e mixinEntries) Len() int {
+	return len(e)
+}
+
+func (e mixinEntries) Swap(i, j int) {
+	e[i], e[j] = e[j], e[i]
+}
+
+func (e mixinEntries) Less(i, j int) bool {
+	return e[i].GetLastUpdated().Before(e[j].GetLastUpdated())
+}
+
+type mixinFile struct {
+	File    string
+	Updated time.Time
+}
+
+type mixinFeed map[string]map[string]*mixinFileset
+
+func Generate(opts GenerateOptions, cxt *context.Context) error {
+	feedTmpl, err := cxt.FileSystem.ReadFile(opts.TemplateFile)
+	if err != nil {
+		return errors.Wrapf(err, "error reading template file at %s", opts.TemplateFile)
+	}
+
+	mixinRegex := regexp.MustCompile(`(.*/)?(.+)/([a-z]+)-([a-z0-9]+)-([a-z0-9]+)(\.exe)?`)
```

Comment:
Do we want to restrict some of these regex fields to only the currently supported values?  (For instance, OS can be only one of `{linux, windows, darwin}`, arch only `amd64`, etc.)

I can also understand wanting to keep it unrestricted with a vision towards supporting other variants...

Reply:
I originally didn't want to be in the business of keeping up to date with our build matrix but honestly, it is not going to change often. So we just need to be aware that this is here. I'll update it and leave a note in the makefile's build matrix that this should be updated too when it's changed.